### ISSUE TITLE
prevent calling onEnter on way back after scrolling down once

### ISSUE THREAD

**Issue by simran66** [brigade/react-waypoint#153]
In my OnEnter callback, I am making an api call to fetch more data. The data fetched is added to the array of elements I have. When I scroll back up, the api call is triggered again as the component is back in the viewport. 

Is there a way to get around this?

**Comment by jmeas** [comment]
There is @simran66 . As you've discovered, `onEnter` is going to be called whenever the user scrolls, whether you need it to or not. What you'll need to do is add some state that can let you know if, when `onEnter` is called, you need to do something.

In this situation, you just need to keep track of when `onEnter` is called for the first time, and only fetch data then.

One way to do this would might look something like this:


```jsx
onWaypointEnter() {
  // If we have already entered before, then we don't do anything.
  if (this._hasEntered) {
    return;
  }

  // Otherwise, we fetch our data and set `_hasEntered` to be true.
  fetchData();
  this._hasEntered = true;
}

render() {
  return (
    <Waypoint onEnter={this.onWaypointEnter}>
  );
}
```

If you're using Lodash, then you could write less code by allowing Lodash to manage the state for you using the [before](https://lodash.com/docs/4.17.4#before) method. For instance,

```jsx
render() {
  return (
    <Waypoint onEnter={_.before(2, this.onWaypointEnter)}>
  );
}
```

Let me know if that helps, @simran66 !

**Comment by simran66** [comment]
Thanks :) I tried this:

```
fetchData(props){
    console.log("position current", props.currentPosition);
    console.log("position previous", props.previousPosition);
    if(props.previousPosition != 'above') {
      console.log("fetch data called");
      console.log("page number before");
      this.setState({page: this.state.page + 1}, function () {
        console.log("page number now is", this.state.page);
        this.getCatalogItems(this.state.categoryId, "anythingForNow", 0, this.state.filters)
          .then((res) => {
            console.log("fetched more data with scroll", res)
          })
      })
    }else{

      console.log("not calling fetch data")
    }
  }
```


Does it make any sense at all ?

**Comment by jmeas** [comment]
> Does it make any sense at all ?

@simran66 , I'm not entirely sure what that code is trying to do. Would you be able to try to write out what the goal is?

> Also, I am not using lodash :/.

That's fine. That was just one option – there are lots of ways to do the same thing!

> How do I destroy it? Could you show me an example?

Sure. Here's a little example:

```js
onWaypointEnter() {
  fetchData();
  this.setState({hasEntered: true});
}

render() {
  var waypoint;

  // Only render the waypoint if it hasn't fired "onEnter" before.
  if (!this.state.hasEntered) {
    waypoint = (<Waypoint onEnter={this.onWaypointEnter}>);
  }
  
  return (
    {waypoint}
    <div>Hello</div>
  );
}
```

Depending on how complex your page is, though, you may need to do something more involved.

**Comment by simran66** [comment]
Okay, let me clean that  up a little bit:

**Comment by jmeas** [comment]
@simran66 , can you clarify if this describes what you're trying to do?

- The webpage loads, and a partial list of items is rendered. Maybe 10 items or so are rendered.
- When the user scrolls to the bottom of the page, another request is made to the server to fetch more of the list (for instance, another 10 items)
- If the user then scrolls to the bottom of the page again, then another set of items is rendered
- And so on until infinity or the end of the list

Does that sound right to you?

**Comment by simran66** [comment]
@jmeas : That's exactly what I am trying to do and my state is maintained through redux

**Comment by jmeas** [comment]
@simran66 , I may not be able to provide working code for you, because this is a more complex thing that you're trying to do. It will likely some tinkering to get the exact behavior you want. I spent a few minutes thinking about how I'd approach implementing this, and I'll write down my thought process here. There are probably many other ways to do it, too – some may be much better than this one!

There will be some code snippets in this comment, but I'm not running them myself, so there may be a few typos. Also, it's been a few weeks since I've written React, so cut me some slack if there are some silly errors :)

First, what I would do is start with the list component rendering the items. You already have this above, but let's start simple. This might look something like...

```js
render() {
  var products = this.props.products;
  return (
    <div>
      {products.map(product => (
        <li key={product.id}>
          {product.name}
        </li>
      )}
    </div>
  );
}
```

If there are 10 products in the store, then this will render out those 10 products. Later, when the data in redux is updated and there are 20, then this will render those 20. So this is a good start.

Next what we need to do is to add the Waypoint to fetch more. We always want this Waypoint to be at the end of the list, since we want to fetch more when the user scrolls down. It looks like you've also got this part down. A first draft of this might look something like:

```js
onWaypointEnter() {
  fetchData();
}

render() {
  var products = this.props.products;
  return (
    <div>
      {products.map(product => (
        <li key={product.id}>
          {product.name}
        </li>
      )}
      <Waypoint onEnter={this.onWaypointEnter}/>
    </div>
  );
}
```

This doesn't work because the fetching of data could take some time. If the user scrolls up and down during that time, the waypoint would make many requests. So we want to remove the Waypoint after the request is made. That might look something like...

```js
constructor(props) {
  super(props);
  // Start off showing the waypoint, so that the user can fetch data when the page loads
  this.state = {
    showWaypoint: true
  };
} 

onWaypointEnter() {
  fetchData();
  this.setState({showWaypoint: false});
}

render() {
  var products = this.props.products;
  return (
    <div>
      {products.map(product => (
        <li key={product.id}>
          {product.name}
        </li>
      )}
      {this.state.showWaypoint && <Waypoint onEnter={this.onWaypointEnter}/>}
    </div>
  );
}
```

Alright, what the above code is intended to do is to turn off the waypoint after a request to fetch data is submitted. So what happens is that the user enters the page, and either immediately or when they scroll down, the Waypoint's `onEnter` will trigger, and data will be fetched.

Upon fetching the data, the state of the component is updated, which will cause the Waypoint to be immediately removed. Therefore, only a single request is made.

This isn't good, though, because we want the user to be able to retrieve data again, just not immediately. We want them to be able to fetch data after the request to fetch data has resolved successfully. One way to track this is to store metadata about the request in redux. In my apps, I keep track of each and every request in the redux store. So, when a request is fired off, somewhere in redux I store that that particular request is in flight. Then, when it succeeds or fails, I once again update the status for that request.

Assuming we have a system like that in place, we can track when the state of our request switches from being in-flight to success. When that happens, we turn the waypoint back on. This can be accomplished in the `componentWillReceiveProps` function.

```js
componentWillReceiveProps(nextProps) {
  if (nextProps.fetchingDataStatus === 'success' && this.props.fetchingDataSuccess === 'pending') {
    this.setState({showWaypoint: true});
  }
}
```

That will get fired anytime the props of the component change. If the status of the request has changed from "pending" to "success," then the waypoint will reappear, because we've updated our state to display it.

This should be a system that at least approximates what you're trying to do. Once the waypoint reappears, the user can scroll back down, encounter the waypoint, and another request will be fired off. This will in turn disable the waypoint until the request succeeds, which will turn back on the waypoint...and so on.

There are some more things to add to this for it to really work. You'd want to add some handling for when the request fails, for instance.

Let me know if that helps!

**Comment by jmeas** [comment]
@simran66 , I'm going to go ahead and close this issue, but feel free to comment here if there's more that I can help with. I know the above comment was a lot of code (and words!), so it might be hard to follow. If you have specific questions about it, or if it just straight up doesn't work, please do leave a comment.

Good luck! 👌
